### PR TITLE
Profiles: more auto-exec triggers

### PR DESCRIPTION
**Pull request by hius07**
New "on wake-up", conditional "on book opening".
Ask to execute a profile (in auto-exec only), show notification on executing.

On book opening:

Closes https://github.com/koreader/koreader/issues/12491.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12651)

### CODE REVIEW

**Comment by hius07** [comment]
I mean a general iterator for any table (branch) size, if it is possible.
Similar to
https://github.com/koreader/koreader/blob/716ed1f19d0c3f49225cd12730155336435d7455/frontend/util.lua#L833-L837

**Comment by mergen3107** [comment]
Is this just a while loop with limited iterations?

**Comment by hius07** [comment]
Yes, but over the keys of a hash-table.
How to represent `t[key1][key2]...[key-n]`?

**Comment by mergen3107** [comment]
Does it accept wildcards?
`t[key*]`?

**Comment by hius07** [comment]
The difficulty is not in a key name but in the variable depth.

**Comment by poire-z** [comment]
(No real idea :/)

**Comment by hius07** [comment]
Anyway, all works well, just an aesthetic question.

**Comment by Frenzie** [comment]
Maybe you could rewrite it a bit but I can't imagine it'd make it any clearer.
```lua
function Profiles:delAutoExecValue(key1, key2, key3, key4) -- deletes empty branch
    local function deleteIfEmpty(keys)
      -- something much more difficult to follow, probably with a backwards for loop
      -- not 100% sure if it could work out :-)
    end

    if key4 then
        deleteIfEmpty{key1, key2, key3, key4}
    elseif key3 then
        deleteIfEmpty{key1, key2, key3}
    elseif key2 then
        deleteIfEmpty{key1, key2}
    else
        self.autoexec[key1] = nil
    end
end
```

I also asked Copilot and it came up with something actually nicer looking than the incomplete thought I suggested. (It looks plausible, but I haven't tested it.)
```lua
function Profiles:delAutoExecValue(...)
    local keys = {...}
    local lastKey = table.remove(keys)
    
    local parent = self.autoexec
    for _, key in ipairs(keys) do
        if parent[key] == nil then
            return
        end
        parent = parent[key]
    end
    
    parent[lastKey] = nil
    
    -- Recursively delete empty branches
    for i = #keys, 1, -1 do
        local key = keys[i]
        if next(parent) == nil then
            parent = self.autoexec
            for j = 1, i - 1 do
                parent = parent[keys[j]]
            end
            parent[key] = nil
        else
            break
        end
    end
end
```

**Comment by hius07** [comment]
Thanks, I tried that idea in the beginning but failed. Now being encouraged by AI :) have succeeded.

It's interesting that reassigning a table to its subtable doesn't affect the original (referenced) table `self.autoexec`:
```Lua
parent = self.autoexec
for j = 1, i - 1 do
    parent = parent[keys[j]]
end
```

File: frontend/util.lua [hunk 1]
```diff
@@ -210,6 +210,52 @@ function util.tableSize(t)
...
+        q = q[key]
+    end
+    q[keys[#keys]] = value
+end
+
+--- Removes a key in a table, removes all empty parent keys.
+---- @param t Lua table
+---- @param ... parent keys, starting from the upper level
+function util.tableRemovePath(t, ...)
+    local keys = { ... }
```
Comment (poire-z) [inline frontend/util.lua:247]: Note sure "Path" is the right word for some generic table function - it's too much "file system" connoted.
May be tableRemoveSubTree ?

Comment (hius07) [reply frontend/util.lua:247]: The purpose is to delete a value of the (last) key. Removing empty subtree is a consequence. So maybe `util.tableRemoveValue` or `util.tableDelValue` (like Get and Set).

Comment (poire-z) [reply frontend/util.lua:247]: > So maybe `util.tableRemoveValue` or `util.tableDelValue` (like Get and Set).

Fine with me (rather full words like Remove or Delete, than just Del).

**Comment by jonnyl2** [comment]
@hius07 Thank you for adding these very useful new features. I'd like to add a suggestion for an additional trigger:

    On book opening -> if book is in collection -> [select collection(s)]

I'd be very interested in using this feature to toggle statistics off on opening of certain books, and toggling them back on on closing. However, 'on book closing' doesn't allow further conditions at the moment. Could the 'on opening' conditions be duplicated here? 

The following suggestion is probably out of scope, because it doesn't concern triggers but actions. However, perhaps it could be considered for a future update: 

    Trigger: On book closing (+conditions). 
    Actions: Reset book, Remove from History.

Sorry if this is not the right place for this, but I couldn't find any relevant discussion in the Issues.

**Comment by hius07** [comment]
Thanks for the ideas, looking interesting.

**Comment by NiLuJe** [comment]
I wouldn't recommend trying to do stuff on document close, because of the lifecycle constraints (i.e., the Document object *will* go poof).

**Comment by NiLuJe** [comment]
That still leaves implementations that may randomly be using `scheduleIn` and/or `nextTick` & co behind your back, which *will* horribly break.

**Comment by hius07** [comment]
This one is ready, more triggers to be added in the next round.

**Comment by Frenzie** [comment]
What does "enter text contained in" do precisely?

**Comment by hius07** [comment]
Enter comma separated patterns to search in the book property to trigger a profile execution.

**Pull request merged**